core/src/memory.rs

```diff
@@ -584,6 +588,7 @@ impl AllocationsContext {
         &mut self,
         pages: WasmPage,
         mem: &mut impl Memory,
+        charge_gas_for_grow: impl FnOnce(WasmPage) -> Result<(), ChargeError>,
```

Comment:
Consider
```suggestion
        counters: &mut impl CountersOwner,
```

Reply:
That was the idea at the very beginning, but the issue is that the `self` has already got shared borrow  and passing it down requires the second shared borrow which is prohibited. Even though if it was possible to do so, the charging requires info about the amount to charge which is calculated based on the structures residing in the `core-processor` whereas the `alloc` sits in the `core`. Hopefully this answers your questions with regards to inlining and using `CountersOwner::charge_gas_if_enough` directly